### PR TITLE
29 add ability to transfer moderator duties

### PR DESCRIPTION
**Pull request by sarashunter**
Tested via postman.  I assumed we would be able to get the voterIds from the fronten, so that we could have more than one voter with the same name.

### CODE REVIEW

File: server/controllers/socket.controller.ts [hunk 1]
```diff
@@ -235,6 +236,52 @@ async function handleOptionSelected(
...
+	);
+	const newVoter = {
+		id: roomData.moderator.id,
+		name: roomData.moderator.name,
+		confidence: null,
+		selection: null,
+	};
+	const updatedVoters = [...votersWithoutNewModerator, newVoter];
+
+	const newModeratorName = roomData.voters.find(
```
Comment (lvl-mattwells) [inline server/controllers/socket.controller.ts:264]: Are we handling the case when the `.find(...)` returns undefined? It shouldn't ever happen, but shouldn't we still handle it?

File: server/controllers/socket.controller.ts [hunk 2]
```diff
@@ -235,6 +236,52 @@ async function handleOptionSelected(
...
+	data: ModeratorChangeEvent,
+): Promise<void> {
+	if (!roomCode) throw new Error("Unable to start voting due to no room code.");
+	const roomData = await lookupRoom(roomCode);
+	if (!roomData) return;
+
+	const votersWithoutNewModerator = roomData.voters.filter(
+		(voter) => voter.id !== data.newModeratorId,
+	);
+	const newVoter = {
```
Comment (lvl-mattwells) [inline server/controllers/socket.controller.ts:256]: Can we specify the type here just for type-checking?

File: server/controllers/socket.controller.ts [hunk 3]
```diff
@@ -235,6 +236,52 @@ async function handleOptionSelected(
...
+		name: roomData.moderator.name,
+		confidence: null,
+		selection: null,
+	};
+	const updatedVoters = [...votersWithoutNewModerator, newVoter];
+
+	const newModeratorName = roomData.voters.find(
+		(voter) => voter.id === data.newModeratorId,
+	).name;
+	const updatedModerator = { id: data.newModeratorId, name: newModeratorName };
```
Comment (lvl-mattwells) [inline server/controllers/socket.controller.ts:267]: Can we specify the type here just for type-checking?

File: server/types/socket.ts [hunk 4]
```diff
@@ -16,6 +16,11 @@ export interface JoinEvent {
 	name: string;
 }
 
+export interface ModeratorChangeEvent {
```
Comment (lvl-mattwells) [inline server/types/socket.ts:19]: Can we add this to the `WebSocketEvent` union type so that the switch-case in `socket.controller.ts` is properly typed

**Review by lvl-mattwells — CHANGES_REQUESTED** [review]

File: server/types/socket.ts (hunk 4 above)
Comment (sarashunter) [reply server/types/socket.ts:19]: Oof yeah. Sorry I just switched laptops and didn't realize that I forgot to turn on the Deno typecheck settings.

File: server/controllers/socket.controller.ts (hunk 1 above)
Comment (sarashunter) [reply server/controllers/socket.controller.ts]: Added more thorough checks.

**Review by harrythree — CHANGES_REQUESTED** [review]

File: server/controllers/socket.controller.ts [hunk 5]
```diff
@@ -282,6 +339,10 @@ export const handleWs = (socket: WebSocket) => {
 					await handleOptionSelected(userId, roomCode, data);
 					break;
 				}
+				case "ModeratorChange": {
+					await handleModeratorChange(roomCode, data);
```
[the remark this answers is not in the document]
Comment (sarashunter) [reply server/controllers/socket.controller.ts:343]: The roomCode doesn't come from the ModeratorChange event.  It's handled the same way as the OptionSelected event above.  I tested it via Postman WS.  I created the room.  Then I opened a WS tab and used a join event.  That user became the moderator by default.  Then I opened another WS tab to join as another user.  I then went back to the first tab and did the moderatorchange event there.  You have to use the actual id of the user not a name.  You can get this from the roomupdate events that go out.

**Review by lvl-mattwells — APPROVED** [review]

**Pull request merged**